### DISCUSSION TITLE
When will a match instance be deleted from the server?

### DISCUSSION THREAD

**Discussion by jakehsiao** [boardgameio/boardgame.io#901] [Q&A] [answered]
After a several hours, when enter the game with the same matchID, the game instance is still here.

Can the server automatically delete unactivated game instance?

**Comment by delucis** [comment, marked as answer]
Currently the server deletes a match when all players use the `leave` endpoint, making the room empty:

https://github.com/boardgameio/boardgame.io/blob/d56728bfc93a654bd12187dfcf1661fe7c78a1b3/src/server/api.ts#L305-L310

In practice this may not happen reliably if a player closes a client before the `leave` request. It would probably be nice to add an option for some time-based clean-up now that matches contain `updatedAt` metadata and an option to disable auto-deletion.

**Comment by jakehsiao** [comment]
@delucis So could you please give me a specific way to make server automatically clean match instances every 24 hours?

**Comment by delucis** [comment]
I’ll leave the precise scheduling logic up to you — you could use a library like [`cron`][cron] or [`node-schedule`][ns] or perhaps just a `setInterval` call — but the basics would be something like the below.

Please be careful with this! If your database connector doesn’t support the `where` option, this would wipe *all* matches. In that case you would need to fetch each match one by one to check `updatedAt` manually.

```js
// Create and run the server as usual.
const { Server } = require('boardgame.io/server');

const db =  new DbConnector();

const server = Server({
  games: [game1, game2, ...],
  db,
});

server.run(8000);

// Define clean-up method.
const day = 24 * 60 * 60 * 1000;
const cleanStaleMatches = async () => {
  const weekAgo = Date.now() - 7 * day;
  // Retrieve matchIDs for matches unchanged for > 1 week.
  const staleMatchIDs = await db.listMatches({
    where: {
      updatedBefore: weekAgo,
    },
  });
  // Delete matches one-by-one. Could also use a queue or parallel system here.
  for (const id of staleMatchIDs) {
    await db.wipe(id);
  }
}

// Schedule clean-up.
setInterval(cleanStaleMatches, day);
```



[cron]: https://www.npmjs.com/package/cron
[ns]: https://www.npmjs.com/package/node-schedule
